File: src/runtime/AuditLogger.ts
import * as fs from 'fs/promises';
import * as path from 'path';
import type { AuditEntry } from '../types.js';

export class AuditLogger {
  private auditDir: string;
  private entries: Map<string, AuditEntry> = new Map();

  constructor(auditDir: string) {
    this.auditDir = auditDir;
  }

  /**
  * Start a new audit entry
   */
  start(nodeId: string, extra: { tempDir: string }): AuditEntry {
    const entry: AuditEntry = {
      id: `${nodeId}-${Date.now()}`,
      timestamp: new Date().toISOString(),
      nodeId,
      phase: 'start',
      execution: {
        tempDir: extra.tempDir,
        duration: 0,
        exitCode: -1
      }
    };

    this.entries.set(entry.id, entry);
    return entry;
  }

  /**
  * Update audit entry phase
   */
  update(entry: AuditEntry, phase: 'validate' | 'execute' | 'retry'): void {
    entry.phase = phase;
    entry.timestamp = new Date().toISOString();
  }

  /**
  * Mark audit entry as complete
   */
  complete(entry: AuditEntry): void {
    entry.phase = 'complete';
    if (entry.execution) {
      entry.execution.duration = Date.now() - new Date(entry.timestamp).getTime();
    }
    this.persist(entry);
  }

  /**
  * Mark audit entry as failed
   */
  error(entry: AuditEntry): void {
    entry.phase = 'error';
    this.persist(entry);
  }

  /**
  * Mark audit entry as skipped
   */
  skipped(entry: AuditEntry): void {
    entry.phase = 'skipped';
    this.persist(entry);
  }

  /**
  * Persist audit entry to disk
   */
  private async persist(entry: AuditEntry): Promise<void> {
    const logFile = path.join(this.auditDir, `${entry.id}.json`);
    await fs.writeFile(logFile, JSON.stringify(entry, null, 2));
  }

  /**
  * Get all audit entries
   */
  getEntries(): AuditEntry[] {
    return Array.from(this.entries.values());
  }
}

File: orc-java/src/main/java/com/orc/runtime/AuditLogger.java
package com.orc.runtime;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.orc.model.AuditEntry;
import com.orc.model.Phase;

import java.io.File;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class AuditLogger {

    private final String auditDir;
    private final Map<String, AuditEntry> entries = new ConcurrentHashMap<>();
    private final ObjectMapper objectMapper = new ObjectMapper();

    public AuditLogger(String auditDir) {
        this.auditDir = auditDir;
        new File(auditDir).mkdirs();
    }

    public AuditEntry start(String nodeId, String tempDir) {
        AuditEntry entry = new AuditEntry(
                nodeId + "-" + System.currentTimeMillis(),
                java.time.Instant.now().toString(),
                nodeId,
                Phase.start,
                null, null,
                new AuditEntry.ExecutionDetail(tempDir, 0, -1, null, null),
                null, null, null);
        entries.put(entry.id(), entry);
        return entry;
    }

    public void update(AuditEntry entry, Phase phase) {
        AuditEntry updated = new AuditEntry(
                entry.id(), java.time.Instant.now().toString(), entry.nodeId(), phase,
                entry.inputs(), entry.outputs(), entry.execution(), entry.retry(), entry.persistedFiles(), entry.error());
        entries.put(entry.id(), updated);
    }

    public void complete(AuditEntry entry) {
        AuditEntry updated = new AuditEntry(
                entry.id(), java.time.Instant.now().toString(), entry.nodeId(), Phase.complete,
                entry.inputs(), entry.outputs(), entry.execution(), entry.retry(), entry.persistedFiles(), entry.error());
        entries.put(entry.id(), updated);
        persist(updated);
    }

    public void error(AuditEntry entry) {
        AuditEntry updated = new AuditEntry(
                entry.id(), java.time.Instant.now().toString(), entry.nodeId(), Phase.error,
                entry.inputs(), entry.outputs(), entry.execution(), entry.retry(), entry.persistedFiles(), entry.error());
        entries.put(entry.id(), updated);
        persist(updated);
    }

    public void skipped(AuditEntry entry) {
        AuditEntry updated = new AuditEntry(
                entry.id(), java.time.Instant.now().toString(), entry.nodeId(), Phase.skipped,
                entry.inputs(), entry.outputs(), entry.execution(), entry.retry(), entry.persistedFiles(), entry.error());
        entries.put(entry.id(), updated);
        persist(updated);
    }

    private void persist(AuditEntry entry) {
        try {
            File logFile = new File(auditDir, entry.id() + ".json");
            objectMapper.writerWithDefaultPrettyPrinter().writeValue(logFile, entry);
        } catch (IOException e) {
            // Log silently, don't fail execution
        }
    }
}
